test/prototype/test_parq.py

```diff
@@ -53,52 +78,184 @@ def forward(self, x):
         quantizer = LSBQuantizer()
-        prox_map = ProxPARQ(anneal_start=0, anneal_end=2)
-        optimizer = QuantOptimizer(base_optimizer, quantizer, prox_map)
-
-        for _ in range(3):
-            x = self.model.example_inputs().to(_DEVICE)
-            out = self.model(x)
-            out.sum().backward()
-            optimizer.step()
+        optimizer = QuantOptimizer(
+            base_optimizer, quantizer, ProxPARQ(anneal_start=0, anneal_end=2)
+        )
+        train_loop(self.model, optimizer, steps=3)
 
         for child in self.model.children():
-            if isinstance(child, torch.nn.Linear):
-                self.assertEqual(child.weight.unique().numel(), 3)
+            if isinstance(child, nn.Linear):
+                self.assertEqual(
+                    child.weight.unique().numel(), quantizer.get_quant_size(b)
+                )
+
+
+class TestUnifTorchaoQuantizer(common_utils.TestCase):
+    def setUp(self):
+        torch.manual_seed(123)
+
+    @staticmethod
+    def int4_torchao_quantizer(b: int = 4, config=None):
+        # based off torchao.quantization.quant_api._int4_weight_only_transform
+        return UnifTorchaoQuantizer(
+            symmetric=False,
+            target_dtype=torch.int32,
+            quant_min=0,
+            quant_max=2**b - 1,
+            eps=1e-6,
+            preserve_zero=False,
+            config=config,
+        )
+
+    def compare_quantized_models(
+        self,
+        model: nn.Module,
+        m_ref: nn.Module,
+        quantizer: UnifTorchaoQuantizer,
+        b: int,
+        group_size: int,
+    ):
+        for n, module in model.named_children():
+            if not _is_linear(module):
+                continue
+
+            # simulate grouping from QuantOptimizer.step
+            p = module.weight
+            original_shape = p.shape
+            p = p.view(-1, group_size)
+
+            q, Q = quantizer.quantize(p, b=b, dim=-1)
+            q = q.view(original_shape)
+
+            # compare to AffineQuantizedTensor instance
+            ref = getattr(m_ref, n).weight.dequantize()
+            self.assertTrue(q.equal(ref))
+
+    @unittest.skipIf(not TORCH_VERSION_AT_LEAST_2_4, "Test only enabled for 2.4+")
+    @unittest.skipIf(_DEVICE == "cpu", "Need GPU available")
+    @common_utils.parametrize("group_size", [32, 256])
+    def test_int4_weight_only(self, group_size: int = 32):
+        model = M(m=512, n=512).to(torch.bfloat16).to(_DEVICE)
+        model.reset_parameters()
+
+        m_ref = copy.deepcopy(model)
+        quantize_(m_ref, Int4WeightOnlyConfig(group_size))
+
+        b = 4
+        quantizer = self.int4_torchao_quantizer()
+        self.compare_quantized_models(model, m_ref, quantizer, b, group_size)
+
+    @unittest.skipIf(not TORCH_VERSION_AT_LEAST_2_6, "Test only enabled for 2.6+")
+    @common_utils.parametrize("b", [2, 3, 4, 8])
+    @common_utils.parametrize("group_size", [32, 512])
```

Comment:
Added parametrized test cases that loop over (int2, int3, int4, int8) data types.